File: app/src/main/java/com/machiav3lli/derdiedas/ui/SettingsFragment.java
package com.machiav3lli.derdiedas.ui;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatDelegate;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import com.machiav3lli.derdiedas.Constants;
import com.machiav3lli.derdiedas.R;
import com.machiav3lli.derdiedas.utils.PrefsUtil;

import static androidx.appcompat.app.AppCompatDelegate.setDefaultNightMode;

public class SettingsFragment extends PreferenceFragmentCompat {
    final static String TAG = Constants.classTag(".PrefsFragment");

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.preferences, rootKey);
        Preference pref;

        pref = findPreference(Constants.PREFS_THEME);
        assert pref != null;
        pref.setOnPreferenceChangeListener((preference, newValue) -> {
            PrefsUtil.setPrefsString(requireContext(), Constants.PREFS_THEME, newValue.toString());
            switch (newValue.toString()) {
                case "light":
                    setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);
                    break;
                case "dark":
                    setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
                    break;
                default:
                    setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM);
            }
            return true;
        });
    }
}

File: app/src/main/java/com/machiav3lli/derdiedas/ui/SettingsFragment.kt
package com.machiav3lli.derdiedas.ui

import android.os.Bundle
import androidx.appcompat.app.AppCompatDelegate
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.machiav3lli.derdiedas.PREFS_LANG
import com.machiav3lli.derdiedas.PREFS_THEME
import com.machiav3lli.derdiedas.R
import com.machiav3lli.derdiedas.utils.appTheme
import com.machiav3lli.derdiedas.utils.language
import com.machiav3lli.derdiedas.utils.restartApp

class SettingsFragment : PreferenceFragmentCompat() {
    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences, rootKey)
        findPreference<Preference>(PREFS_THEME)?.onPreferenceChangeListener =
            Preference.OnPreferenceChangeListener { _: Preference?, newValue: Any ->
                requireContext().appTheme = newValue.toString()
                when (newValue.toString()) {
                    "light" -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
                    "dark" -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
                    else -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
                }
                true
            }

        findPreference<ListPreference>(PREFS_LANG)?.apply {
            val oldLang = value
            onPreferenceChangeListener =
                Preference.OnPreferenceChangeListener { _: Preference?, newValue: Any ->
                    val changed = oldLang != newValue.toString()
                    if (changed) {
                        requireContext().language = newValue.toString()
                        requireContext().restartApp()
                    }
                    changed
                }
        }
    }
}